### PR TITLE
feat(rome_js_analyze): useIframeTitle

### LINKED ISSUE #3945: useIframeTitle
https://github.com/rome/tools/blob/archived-js/website/src/docs/lint/rules/a11y/useIframeTitle.md

### PR DESCRIPTION
**Pull request by unvalley**
## Summary

Closes #3945 

- archived-js: [archived-js/website/src/docs/lint/rules/a11y/useIframeTitle.md](https://github.com/rome/tools/blob/archived-js/website/src/docs/lint/rules/a11y/useIframeTitle.md?rgh-link-date=2022-12-05T09%3A12%3A23Z)
- eslint-plugin-jsx-a11y: https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/main/docs/rules/iframe-has-title.md

## Test Plan

`cargo test -p rome_js_analyze -- use_iframe_title`

## Documentation

- [x] The PR requires documentation
	- and I executed `cargo lintdoc` 
- ~[ ] I will create a new PR to update the documentation~

### CODE REVIEW

File: crates/rome_js_analyze/src/semantic_analyzers/nursery/use_iframe_title.rs [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+impl Rule for UseIframeTitle {
+    type Query = Ast<AnyJsxElement>;
+    type State = UseIframeTitleState;
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+
+        let is_iframe = matches!(node.name_value_token()?.text_trimmed(), "iframe");
```
Comment (xunilrj) [inline crates/rome_js_analyze/src/semantic_analyzers/nursery/use_iframe_title.rs:88]: You don't need `matches` here. You can do

```rust
if node.name_value_token()?.text_trimmed() != "iframe" {
    return None;
}
```

directly.

File: crates/rome_js_analyze/src/semantic_analyzers/nursery/use_iframe_title.rs [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        if !is_iframe {
+            return None;
+        }
+
+        if node.attributes().is_empty() {
```
Comment (xunilrj) [inline crates/rome_js_analyze/src/semantic_analyzers/nursery/use_iframe_title.rs:93]: If the diagnostic message is the same in both cases you can merge them like:

```rust
let Some(title_attribute) = node.find_attribute_by_name("title") else {
    return Some(UseIframeTitleState { node: node.clone() });
}
let attribute_value = title_attribute.initializer()?.value().ok()?;

match attribute_value  {
    JsxString(str) => is_valid_string(str),
    JsxExpressionAttributeValue(expr) => is_valid_expression(expr),
    AnyJsxTag => return Some(UseIframeTitleState { node: node.clone() });
}
```

If `node.attributes().is_empty()`, `find_attribute_by_name` will fail fast enough already.

File: crates/rome_js_analyze/src/semantic_analyzers/nursery/use_iframe_title.rs (hunk 1 above)
Comment (unvalley) [reply crates/rome_js_analyze/src/semantic_analyzers/nursery/use_iframe_title.rs]: done at [31d415b](https://github.com/rome/tools/pull/4067/commits/31d415bda36c647def9c564efc842201945833a5)

File: crates/rome_js_analyze/src/semantic_analyzers/nursery/use_iframe_title.rs (hunk 2 above)
Comment (unvalley) [reply crates/rome_js_analyze/src/semantic_analyzers/nursery/use_iframe_title.rs:92]: fixed at [`e26c383` (#4067)](https://github.com/rome/tools/pull/4067/commits/e26c383d26a164c0f94c271b88deb886845ee226)

File: crates/rome_js_analyze/src/semantic_analyzers/nursery/use_iframe_title.rs [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+            }
+            AnyJsxAttributeValue::AnyJsxTag(_) => Some(UseIframeTitleState { node: node.clone() }),
+        }
+    }
+
+    fn diagnostic(_ctx: &RuleContext<Self>, state: &Self::State) -> Option<RuleDiagnostic> {
+        Some(
+            RuleDiagnostic::new(
+                rule_category!(),
+                state.node.syntax().text_trimmed_range(),
```
Comment (leops) [inline crates/rome_js_analyze/src/semantic_analyzers/nursery/use_iframe_title.rs:134]: It seems like `state.node` is always set to the query node, in this case you don't need to clone it into the state and can simply retrieve it in the `diagnostic` method by calling `ctx.query()` (then you can just use `()` as the signal type)

File: crates/rome_js_analyze/tests/specs/nursery/useIframeTitle/invalid.jsx.snap [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+```
+
+```
+invalid.jsx:5:2 lint/nursery/useIframeTitle ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
+
+  ! Provide a title attribute when using iframe elements.
+  
+    3 │ 	<iframe />
+    4 │ 	<iframe></iframe>
+  > 5 │ 	<iframe {...props} />
```
Comment (ematipico) [inline crates/rome_js_analyze/tests/specs/nursery/useIframeTitle/invalid.jsx.snap:66]: This case should not trigger the rule. `props` **might** contain `title`, we don't know that.

**Review by ematipico — COMMENTED** [review]
Once we address the feedback, I think we can merge the PR

File: crates/rome_js_analyze/tests/specs/nursery/useIframeTitle/invalid.jsx.snap (hunk 4 above)
Comment (unvalley) [reply crates/rome_js_analyze/tests/specs/nursery/useIframeTitle/invalid.jsx.snap]: I've fixed it at [8100a2c](https://github.com/rome/tools/pull/4067/commits/8100a2c9ed0a2bb95135190ebb3e14c8d09cd697).
imho, this behavior will be different from ESLint.

File: crates/rome_js_analyze/src/semantic_analyzers/nursery/use_iframe_title.rs (hunk 3 above)
Comment (unvalley) [reply crates/rome_js_analyze/src/semantic_analyzers/nursery/use_iframe_title.rs]: Thanks, I've done at [8100a2c](https://github.com/rome/tools/pull/4067/commits/8100a2c9ed0a2bb95135190ebb3e14c8d09cd697).

**Review by leops — APPROVED** [review]

**Pull request merged**